### PR TITLE
Add constant deduplication pass in flat_ir

### PR DESCRIPTION
**Pull request by parthchadha**

### CODE REVIEW

File: tripy/tests/flat_ir/passes/test_constant_deduplication.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    # Verify that one constant op was removed
+    assert len(flat_ir.ops) == original_op_count - 1, "One constant op should have been removed"
+
+    # Verify that only two ConstantOps remain
+    constant_ops = [op for op in flat_ir.ops if isinstance(op, ConstantOp)]
+    assert len(constant_ops) == 2, "There should be only two ConstantOps after deduplication"
+
+    # Verify that the remaining ConstantOps have different data
+    constant_data = [tuple(op.data) for op in constant_ops]
+    assert len(set(constant_data)) == 2, "The remaining ConstantOps should have different data"
```
Comment (pranavm-nvidia) [inline tripy/tests/flat_ir/passes/test_constant_deduplication.py:72]: Is it overkill to also make sure the data is unchanged?

File: tripy/tests/frontend/trace/ops/test_slice.py [hunk 2]
```diff
@@ -81,18 +81,8 @@ def test_invalid_index(self):
...
             tp.TripyException,
-            # Looking three instance of the following:
-            # |             a[5, 3].eval()
-            # |               ^
-            #
-            # and three instances of the following:
-            # |             a[5, 3].eval()
-            # |                  ^
-            match=(3 * first_dim_regex + 3 * second_dim_regex),
-            has_stack_info_for=[a],
```
Comment (pranavm-nvidia) [inline tripy/tests/frontend/trace/ops/test_slice.py:96]: Why are we removing this check?

Comment (parthchadha) [reply tripy/tests/frontend/trace/ops/test_slice.py:96]: The error seems to be throw earlier now

File: tripy/tests/flat_ir/passes/test_constant_deduplication.py (hunk 1 above)
Comment (parthchadha) [reply tripy/tests/flat_ir/passes/test_constant_deduplication.py:72]: Overkill in unit test?

File: tripy/tripy/flat_ir/flat_ir.py [hunk 3]
```diff
@@ -48,6 +63,21 @@ def __str__(self):
...
+    def optimize(self):
+        """
+        Runs all optimization passes attached to this FlatIR class.
+        """
+        for pass_class in self._passes:
+            pass_instance = pass_class()
+            pass_instance.run(self)
+
+    @classmethod
+    def attach_pass(cls, pass_class):
```
Comment (pranavm-nvidia) [inline tripy/tripy/flat_ir/flat_ir.py:75]: Not sure I understand the utility of this method. Why not specify passes directly in the `_passes` member?

Comment (pranavm-nvidia) [reply tripy/tripy/flat_ir/flat_ir.py:75]: Ah, I see, we want to be able to add passes out of line. Could we just have a metaclass that does that automatically when inheriting from `FlatIRPass`?

File: tripy/tripy/flat_ir/passes/constant_deduplicate.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    def _get_constant_key(self, op):
+        from mlir_tensorrt.runtime._mlir_libs._api import MemRefValue
+        from tripy.utils.utils import list_to_tuple
+
+        if isinstance(op.data, MemRefValue):
+            from tripy.backend.mlir.memref import tolist
+
+            l = tolist(op.data)
+            data = tuple(list_to_tuple(l if isinstance(l, List) else [l]))
```
Comment (pranavm-nvidia) [inline tripy/tripy/flat_ir/passes/constant_deduplicate.py:55]: `tuple(list_to_tuple(...))` is probably redundant in this case

File: tripy/tests/flat_ir/passes/test_constant_deduplication.py (hunk 1 above)
Comment (pranavm-nvidia) [reply tripy/tests/flat_ir/passes/test_constant_deduplication.py:72]: I meant should we also check `op.data == [1, 2]`?

File: tripy/tests/frontend/trace/ops/test_slice.py (hunk 2 above)
Comment (pranavm-nvidia) [reply tripy/tests/frontend/trace/ops/test_slice.py:96]: Shouldn't this throw from the trace op? Why does the FlatIR pass affect it?

File: tripy/tripy/flat_ir/flat_ir.py [hunk 5]
```diff
@@ -167,12 +172,31 @@ def register_tensor(self, tensor: "FlatIRTensor") -> "FlatIRTensor":
...
         return tensor
 
+    def _get_constant_key(self, op):
+        from mlir_tensorrt.runtime._mlir_libs._api import MemRefValue
+        from tripy.utils.utils import list_to_tuple
+
+        if isinstance(op.data, MemRefValue):
+            from tripy.backend.mlir.memref import tolist
+
+            l = tolist(op.data)
```
Comment (jhalakpatel) [inline tripy/tripy/flat_ir/flat_ir.py:182]: can we use `memref.ptr` as a cache key here?

File: tripy/tripy/flat_ir/flat_ir.py [hunk 6]
```diff
@@ -186,20 +210,35 @@ def register_tensor_and_collect_ops(tensor, seen_tensors):
...
             register_tensor_and_collect_ops(start_tensor, seen_tensors)
 
+        # Apply tensor replacements
+        for op in new_ops:
+            op.inputs = [self.tensor_replacements.get(inp.name, inp) for inp in op.inputs]
+
         # Rebind the ops to tensors from this FlatIR
-        for op in ops:
+        for op in new_ops:
             op.inputs = [self.register_tensor(inp) for inp in op.inputs]
```
Comment (jhalakpatel) [inline tripy/tripy/flat_ir/flat_ir.py:239]: is registering a tensor twice a side effect free?

**Review by pranavm-nvidia — APPROVED** [review]

File: tripy/tripy/flat_ir/flat_ir.py (hunk 6 above)
Comment (parthchadha) [reply tripy/tripy/flat_ir/flat_ir.py:243]: Where are we registering twice?

**Pull request merged**